### ISSUE TITLE
Question on action bound

### ISSUE THREAD

**Issue by dbsxdbsx** [timoklein/car_racer#8]
Game of openAI "CarRacing-v0" has  actions of 3 dimensions, relatively in rage [-1,1], [0,1] and [0,1].
But from your code, There is only code that bounds action and log_Prob into range [-1,1], no code for in range [0,1].
Did you ignore it?

**Comment by timoklein** [comment]
Hey there,
it's been a long time but I think we ignored it. 

You can see for instance here in the CarRacer repo (https://github.com/openai/gym/blob/master/gym/envs/box2d/car_dynamics.py#L138) that the gas is automatically clipped anyway.
If look further below in the .step() function of the Car (https://github.com/openai/gym/blob/master/gym/envs/box2d/car_dynamics.py#L204), you can see that braking only gets applied if it is > 0.

In either case, I would advise you to look for a more updated repository. This was mine (and the others) first real RL project and we made lots of mistakes :) I also do not really maintain it any more.

Hope that helps!

**Comment by dbsxdbsx** [comment]
@timoklein , thanks for your reply. But, the question I raised is not aiming on a specific game environment.  My core question is how to transfer an (log_prob of ) action to a specific aciton range, say target action bound may be in [2,10]. From section `Enforcing Action Bounds` of paper [SAC](https://arxiv.org/pdf/1812.05905.pdf), they only bound the action into [-1,1], and gives corresponding `log_prob` formula for the new action space, which is [-1,1].

Anyway, I am wondering how to generate the action sampled (and corresponding log_prob) from SAC to any target action space--- And, does it necessary to modify formula of `log_prob` for a specific action bound?  

I also posted in pytorch forum [here](https://discuss.pytorch.org/t/activation-function-and-distribution-for-action-space-between-0-and-1/84887).

Personally, I notice that clipping is not enough to solve it. Just like the target action bound [2,10], it is not feasible to transfer action from range[-1,1] to [2,10] by clipping. But by formula [here](https://stackoverflow.com/questions/49911206/how-to-restrict-output-of-a-neural-net-to-a-specific-range), it is possible, but what about the corresponding `log_prob`?

**Comment by timoklein** [comment]
Ah okay, I think I got you now. I also read the pytorch forum thread. Here's my thoughts on this:

- A transformed Beta distribution policy is an option. See for example in [this paper](https://arxiv.org/pdf/1805.09613.pdf). However, I haven't personally gotten Beta distributions to work. The training is always very unstable.
- Have the network output actions in [-1, 1] and transform to the desired range as part of the environment. See [here](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html).

In general, doing a location-scale transformation of your clamped actions in [-1, 1] should not affect the SAC entropy correction formula. To see why, consider your transformation of the type l + c\*a, where a = tanh(u) just like in the SAC paper (l being a location term, c being a scaling factor).
![](https://discuss.pytorch.org/uploads/default/original/3X/b/d/bd5629b56fefadc5c22e2b775f6f69e9a7acda0d.png)
The term in the log behind the sum comes from the derivate of your tanh transformation. If you instead take the derivative of l + c\*tanh(u), the term l will immediately fall away and you will have
![image](https://user-images.githubusercontent.com/36668911/135235521-986a4d26-5f9b-4be2-9f01-f466c03841fe.png)
as the derivative of the tanh-function applied element-wise to your actions.  
You then want to take the log of the following density
![image](https://user-images.githubusercontent.com/36668911/135235890-b473f643-65b1-4778-81b5-761038fbc836.png)
The scaling term c will persist, but due to the log you can move it out of the sum.  
![image](https://user-images.githubusercontent.com/36668911/135236026-10a0d9d1-d3cf-42b8-bf30-edb036b4abd9.png)
Here D is the dimensionality of your action space.   
Now when you want to perform backprop, the term D log c only depends on the constants D and c. These are irrelevant for the derivative.  
**What does this mean for you?**
You can use 6+4\*tanh(u) to transform your actions between [2, 10] _without having to change the entropy correction formula_.

Source: This has been part of my master's thesis.

Does that answer and solve your question?

**Comment by dbsxdbsx** [comment]
@timoklein , thanks  for your quite concrete answer. Because I am not familiar to the deriavtive when it comes with  linear algebra, I raised this question. Therefore, as long as the derivative process is correct, the answer is yes, it indeed solves my question ^_^. 

And back to the "CarRacing-v0" game of your code, to make it more precise, I think there only need a little modification with method ` sample(self, state):` of class `GaussianPolicy` in `sac/model.py`, where only need to do some modification to let the last dimension of action fall into spectrum [0,1](the correct range for the last action dimension). Also according to your answer, there is no need to modify the `entropy` part of code, as long as it is only used to calculate derivative.

**Comment by timoklein** [comment]
@dbsxdbsx 
Yes, you are right about the `sample` method.

About the formula: It's not explained in detail in the SAC paper, so it's hard to understand just from the paper itself...

I will also post my answer to the Pytorch forums in case somebody looks up your question in the future and wonders what happened.

**Comment by dbsxdbsx** [comment]
@timoklein .Just another question. After reading several open source SAC code including yours, I notice that the output of `deviation` along with `mean` is usually like this form:
```python 
LOG_SIG_MAX = 2
LOG_SIG_MIN = -20

log_std = self.log_std_linear(x)
log_std = torch.clamp(log_std, min=LOG_SIG_MIN, max=LOG_SIG_MAX)
...
mean, log_std = self.forward(state)
std = log_std.exp()
normal = Normal(mean, std)
```
I wonder why clipping log_std in this range [-20,2]. Are these just 2 empirical numbers?
And in my opinion, since the output of  `log_std_linear` cannot be controlled, I would like to change code into this form:
```python 
log_std = self.log_std_linear(x)
...
mean, log_std = self.forward(state)
std = torch.sigmoid(log_std) # make it a positive number in range [0,1], which is more simple and symmetric 
normal = Normal(mean, std)
```
Do you think it a constructive idea?

**Comment by timoklein** [comment]
@dbsxdbsx 
I would strongly advise against using `torch.sigmoid` for the standard deviation. Yes, it will technically be a valid standard deviation. But you will limit how expressive your policy can be. Look at the picture below for instance (showing a tanh-transformed Normal distribution which is rescaled between [-5, 5]): The density of your policy when using the `torch.sigmoid` will not be able to take the yellow shape since that requires a standard deviation > 1.
![Screenshot from 2021-09-30 07-57-59](https://user-images.githubusercontent.com/36668911/135395676-5f7c38c7-caed-43d8-876a-61d6d09d74d2.png)
Learning the log standard deviation and then exponentiating it serves two purposes:

1. You can learn values that are both positive and negative. This helps with training stability.
2. Exponentiating any value will always be >0 and as such produce a valid standard deviation.

Clipping between [-20, 2] helps with numerical stability as well. The exact values depend on the implementation. See e.g.
https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/sac/policies.py vs https://github.com/denisyarats/pytorch_sac/blob/master/config/agent/sac.yaml .  
It generates a distribution that isn't too extreme by preventing the standard deviation from reaching exactly 0 (distribution has collapsed) or from being too high (only choosing extremes).

**Comment by dbsxdbsx** [comment]
@timoklein , You are right that `sigmoid` has limited range in [0,1].  But  also when clipping between [-20, 2] , the output of `e^deviation ` has a limited range in almost in (0,7.389].  If the output of `mean` is quite large or small, say 1000 or 0.001, then I think the corresponding  deviation should also be rescaled to match the magnitude of `mean`. How  do you think?

**Comment by timoklein** [comment]
If you want to learn large values I would advise that you use the transformation outlined in https://github.com/timoklein/car_racer/issues/8#issuecomment-929974828 with large l and c.  
Directly learning a mean of 1000 for instance will make your training very unstable. Your network will need to learn large weights, which will lead to gradients containing large values and this in turn will increase the variance of the training process.

**Comment by dbsxdbsx** [comment]
@timoklein , ok,  thanks.

**Comment by dbsxdbsx** [comment]
> @dbsxdbsx I would strongly advise against using `torch.sigmoid` for the standard deviation. Yes, it will technically be a valid standard deviation. But you will limit how expressive your policy can be. Look at the picture below for instance (showing a tanh-transformed Normal distribution which is rescaled between [-5, 5]): The density of your policy when using the `torch.sigmoid` will not be able to take the yellow shape since that requires a standard deviation > 1. ![Screenshot from 2021-09-30 07-57-59](https://user-images.githubusercontent.com/36668911/135395676-5f7c38c7-caed-43d8-876a-61d6d09d74d2.png) Learning the log standard deviation and then exponentiating it serves two purposes:
> 
> 1. You can learn values that are both positive and negative. This helps with training stability.
> 2. Exponentiating any value will always be >0 and as such produce a valid standard deviation.
> 
> Clipping between [-20, 2] helps with numerical stability as well. The exact values depend on the implementation. See e.g. https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/sac/policies.py vs https://github.com/denisyarats/pytorch_sac/blob/master/config/agent/sac.yaml . It generates a distribution that isn't too extreme by preventing the standard deviation from reaching exactly 0 (distribution has collapsed) or from being too high (only choosing extremes).

@timoklein ，I  am still a little confused on the `mean` and `std` (standard deviation).
The y axis of  figure you plotted is about density on entropy, right? If that is the case, I followed you that to make policy more expressive, std needs to be large to a certain degree, given a specific mean under normal distribution.  But I guess it also make sense to not put mean and std too far away.
What I mean is that, to make a policy expressive and stable enough. Both mean and std should be considered together. 

So, I think rescale `mean_layer` with `tanh` to put it into [-1,1] and still give `std_layer` an sigmoid to put it into [0,1] a proper way to both make policy stable ande expressive. Code changed from this:
```python
log_std = self.log_std_linear(x)
...
mean, log_std = self.forward(state)
std = torch.sigmoid(log_std) # make it a positive number in range [0,1], which is more simple and symmetric 
normal = Normal(mean, std)
```
to this:
```python
log_std = self.log_std_linear(x)
...
mean, log_std = self.forward(state)
mean = mean.tanh()
std = torch.sigmoid(log_std) # make it a positive number in range [0,1], which is more simple and symmetric 
normal = Normal(mean, std)
```
Do I miss something?

**Comment by timoklein** [comment]
@dbsxdbsx ,  Sorry for the late reply, been busy...

> The y axis of figure you plotted is about density on entropy, right? 

It is the density value of the transformed distribution at that point. I have obtained it by:

1. Generate points with `linspace` on the interval $[-5, 5]~.$
2. Create a tanh-squashed Normal distribution with the given mean and standard deviation.
3. Calculate the density value with `.log_prob()` and `.exp()`.

> So, I think rescale mean_layer with tanh to put it into [-1,1]

It's fine to do this but you might have to re-scale the `mean.tanh()` value depending on your action space or if you want to obtain values across a certain range when using the tanh-squashed Normal.
See page 14 (Appendix) of the [Dreamer paper](https://arxiv.org/pdf/1912.01603.pdf) for some info on this.

> give std_layer an sigmoid to put it into [0,1] a proper way to both make policy stable ande expressive.

I still think that especially when you're using the SAC squashed distribution that using a sigmoid for the standard deviation can hinder performance. This happened to me because the algorithm wasn't able to fully explore extreme actions at the start. In the end though you will obtain a valid standard deviation, so I guess you can just try it and see what performs best.
